turbopack/crates/turbopack-core/src/module_graph/chunk_group_info.rs

```diff
@@ -0,0 +1,345 @@
+use std::{
+    collections::{hash_map::Entry, BinaryHeap, HashMap},
+    hash::{DefaultHasher, Hash, Hasher},
+    ops::{Deref, DerefMut},
+};
+
+use anyhow::Result;
+use either::Either;
+use petgraph::graph::{DiGraph, EdgeIndex, NodeIndex};
+use roaring::RoaringBitmap;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use turbo_rcstr::RcStr;
+use turbo_tasks::{
+    debug::ValueDebugFormat, trace::TraceRawVcs, NonLocalValue, ReadRef, ResolvedVc,
+    TryJoinIterExt, ValueToString, Vc,
+};
+
+use crate::{
+    chunk::ChunkingType,
+    module::Module,
+    module_graph::{
+        GraphNodeIndex, GraphTraversalAction, ModuleGraph, SingleModuleGraphModuleNode,
+        SingleModuleGraphNode,
+    },
+};
+
+#[derive(Clone, Debug, Default, PartialEq, TraceRawVcs, ValueDebugFormat)]
+pub struct RoaringBitmapWrapper(#[turbo_tasks(trace_ignore)] pub RoaringBitmap);
+unsafe impl NonLocalValue for RoaringBitmapWrapper {}
+
+// RoaringBitmap doesn't impl Eq, not sure why: https://github.com/RoaringBitmap/roaring-rs/issues/302
+impl Eq for RoaringBitmapWrapper {}
+
+impl Deref for RoaringBitmapWrapper {
+    type Target = RoaringBitmap;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+impl DerefMut for RoaringBitmapWrapper {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl Serialize for RoaringBitmapWrapper {
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        use serde::ser::Error;
+        let mut bytes = vec![];
+        self.0.serialize_into(&mut bytes).map_err(Error::custom)?;
+        serializer.serialize_bytes(bytes.as_slice())
+    }
+}
+
+impl<'de> Deserialize<'de> for RoaringBitmapWrapper {
+    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
+        use serde::de::Error;
+        let bytes = <&[u8]>::deserialize(deserializer)?;
+        let map = RoaringBitmap::deserialize_unchecked_from(bytes).map_err(Error::custom)?;
+        Ok(RoaringBitmapWrapper(map))
+    }
+}
+
+#[turbo_tasks::value(transparent)]
+pub struct ChunkGroupInfo(HashMap<ResolvedVc<Box<dyn Module>>, RoaringBitmapWrapper>);
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+enum ChunkGroup {
+    /// e.g. a page
+    Entry(ReadRef<RcStr>),
+    /// a module that has an incoming async edge
+    Async(ReadRef<RcStr>),
+    /// a module with a incoming non-merged isolated edge
+    Isolated(ReadRef<RcStr>),
```

Comment:
`ReadRef<RcStr>` is pretty expensive to Hash, so a `ResolvedVc<Module>` would be better

Reply:
I had that earlier.
All of that `ReadRef<RcStr>` and `Box<ChunkGroup>` is temporary to be able to debug chunk groups and why (or not) they are created at the right places.